### PR TITLE
Registration Refactor

### PR DESCRIPTION
**Pull request by karimomino**
_**NOTE: This pull request focused on the registration flow and thus only USER,NICK, and PASS commands have been implemented.**_

**Description:**
Added command execution parsing, which will check if a command sent in is valid or not,
if valid it will execute the correct command. So far NICK | PASS | USER commands have been implemented. This completes the registration of client.
Once client is registered, the server converts the client from PreClient -> Client object and send connecting messages to the client. for this to work
accepting connection only happens in PreClient and closing the socket only happens in Client destructor.

**Changes:**
	- accept connection in PreClient not in AClient constructor
	- added 2 utility functions, splitDelim and trim
	- execCommand function added that will parse the rawCommand and execute the correct command
	- NICK | USER | PASS commands implemented
	- converting PreClient to Client upon completed registration
	- Closing the client fd happens only in CLient destructor now not in PreClient destructor also.
	- Added AClient copy constructor for the conversion from PreClient -> Client

### CODE REVIEW

File: src/utils.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+		if ( std::isspace(*i) || !std::isprint(*i) ) {
+			str.erase(i);
+			i = str.end();
+		}
+		else
+			break;
+	}
+}
+
+std::vector<std::string> splitDelim( std::string str , std::string delim) {
```
Comment (ghamry03) [inline src/utils.cpp:25]: this function can be cleaner if you remove the lines after the if condition. like this
```C++
std::vector<std::string> utils::split( std::string str, const std::string& delm ) {
	size_t pos = 0;
	std::string token;
    std::vector<std::string> list;
	while ( ( pos = str.find( delm ) ) != std::string::npos ) {
		token = str.substr(0, pos);
		list.push_back( token );
		str.erase(0, pos + delm.length());
	}
    return ( list );
}
```

File: src/utils.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "Server.hpp"
+
+void trim( std::string& str ) {
```
Comment (ghamry03) [inline src/utils.cpp:3]: You can use the `find_first_not_of()` and `find_last_not_of()` to find where does the actual string starts so the code looks more simple.
```C++
void trim( std::string& str ) {
    const std::string set = " "; // add all the unwanted chars
    size_t start = str.find_first_not_of( set );
    size_t end = str.find_last_not_of( set );
    str = str.substr( start, end - start + 1 );
}
```

**Review by ghamry03 — APPROVED** [review]

File: src/utils.cpp (hunk 2 above)
Comment (karimomino) [reply src/utils.cpp:3]: did not think of that, will be implementing this in the next pull request

File: src/utils.cpp (hunk 1 above)
Comment (karimomino) [reply src/utils.cpp:25]: will be looking into this and possibly implementing it for the next PR

**Pull request merged**